lib/pbench/server/api/__init__.py

```diff
@@ -75,6 +79,103 @@ def create_app():
+    response.headers.add("Access-Control-Allow-Headers", "*")
+    response.headers.add("Access-Control-Allow-Methods", "*")
+    return response
+
+
+def _corsify_actual_response(response):
+    response.headers.add("Access-Control-Allow-Origin", "*")
+    return response
+
+
+class GraphQL(Resource):
+    def post(self):
+        global app
+        graphQL = app.config_graphql.get("server")
+        json_data = request.get_json()
+
+        if not json_data:
+            app.logger.debug("Invalid query object")
+            abort(400, message="Invalid query object in request")
+
+        try:
+            # query GraphQL
+            response = requests.post(graphQL, json=json_data)
+
+            # construct response object
+            response = make_response(response.text)
+            response = _corsify_actual_response(response)
+        except Exception:
+            app.logger.exception("There was something wrong with the POST request.")
+            abort(500, message="There was something wrong with your graphql request")
+        response.status_code = 200
+        return response
+
+    # From the client side, an OPTIONS request is initiated before every POST to request CORS.
+    # On the server side, the server must specify the response headers to grant CORS accordingly.
+
+    def options(self):
+        try:
+            response = _build_cors_preflight_response()
+        except Exception:
+            app.logger.exception("There was something wrong with the OPTIONS request.")
+            abort(400, message="There was something wrong with your request")
+        response.status_code = 200
+        return response
+
+
+class Elasticsearch(Resource):
+    def post(self):
+        global app
+        elasticsearch = app.config_elasticsearch.get("server")
+        json_data = request.get_json()
+        if not json_data:
+            app.logger.debug("Invalid json object")
+            abort(400, message="Invalid json object in request")
+
+        if not json_data["indices"]:
+            app.logger.debug("Missing path")
+            abort(400, message="Missing path in request")
+
+        try:
+            # query ElasticSearch
+            json_data = request.get_json()
+            if "params" in json_data:
+                url = f"{elasticsearch}{json_data['indices']}?{json_data['params']}"
+            else:
+                url = f"{elasticsearch}{json_data['indices']}"
+
+            if "payload" in json_data:
+                response = requests.post(url, json=json_data["payload"])
+            else:
+                response = requests.get(url)
+
+            # construct response object
+            response = make_response(response.text)
+            response = _corsify_actual_response(response)
+        except Exception:
+            app.logger.exception("There was something wrong with the POST request.")
+            abort(500, message="There was something wrong with your download request")
```

Comment:
Especially since we're constructing the URI here, and deciding whether to POST or GET, I think we really need more information here to be sure we understand what went wrong if something happens. Maybe at least that we "POST to url xyzzy" or "GET url xyzzy". And the Elasticsearch HTTP response would be great, too.

Reply:
Good point, I can use this idea in the new PR to make logging more informative #1769